src/algebra/ring/projections.lean

```diff
@@ -0,0 +1,56 @@
+/-
+Copyright (c) 2022 Christopher Hoskin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Christopher Hoskin
+-/
+import algebra.ring.basic
+
+/-!
+# Idempotents
+
+This file defines idempotents for an arbitary multiplication and proves some basic results:
+
+* In a semigroup, the product of two commuting idempotents is an idempotent;
+* In a (non-associative) ring, p is an idempotent if and only if 1-p is an idempotent.
+
+## Tags
+
+projection, idempotent
+-/
+
+variables {M : Type*} [has_mul M]
+
+/--
+An element `p` is said to be idempotent if `p * p = p`
+-/
+def is_idempotent_elem (p : M) : Prop := p*p = p
+
+lemma idempotent_elem_def {p : M} (h : is_idempotent_elem p) : p*p = p := h
+
+namespace is_idempotent_elem
+
+variables {S : Type*} [semigroup S]
+
+lemma mul_of_commute {p q : S} (h : commute p q) (h₁ : is_idempotent_elem p)
+  (h₂ : is_idempotent_elem q) : is_idempotent_elem (p * q)  :=
+begin
+  rw is_idempotent_elem at h₁ h₂,
+  rw [commute, semiconj_by] at h,
+  rw [is_idempotent_elem, mul_assoc, ← mul_assoc q, ←h, mul_assoc p, h₂, ← mul_assoc, h₁],
```

Comment:
With the above rename suggestion, I think this can become
```suggestion
  rw [is_idempotent_elem, mul_assoc, ← mul_assoc q, ←h.eq, mul_assoc p, h₂.eq, ← mul_assoc, h₁.eq],
```